base/src/main/java/io/spine/validate/ExternalConstraintOptions.java

```diff
@@ -0,0 +1,162 @@
+/*
+ * Copyright 2019, TeamDev. All rights reserved.
+ *
+ * Redistribution and use in source and/or binary forms, with or without
+ * modification, must retain the above copyright notice and the following
+ * disclaimer.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+package io.spine.validate;
+
+import com.google.common.collect.ImmutableMap;
+import com.google.protobuf.DescriptorProtos.FieldOptions;
+import com.google.protobuf.Descriptors.Descriptor;
+import com.google.protobuf.Descriptors.FieldDescriptor;
+import com.google.protobuf.GeneratedMessage.GeneratedExtension;
+import io.spine.annotation.Internal;
+import io.spine.code.proto.FieldContext;
+import io.spine.logging.Logging;
+import org.slf4j.Logger;
+
+import java.io.Serializable;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Optional;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+import static com.google.common.collect.Maps.newHashMapWithExpectedSize;
+
+/**
+ * Provides option value for a field mentioned in an {@link ExternalMessageConstraint}.
+ */
+final class ExternalConstraintOptions implements Serializable {
+
+    private static final long serialVersionUID = 0L;
+
+    /**
+     * A map from a field context to the options extracted from an external message constraint.
+     */
+    private final ImmutableMap<FieldContext, FieldOptions> options;
+
+    private ExternalConstraintOptions() {
+        this(new Builder().build());
+    }
+
+    private ExternalConstraintOptions(ImmutableMap<FieldContext, FieldOptions> options) {
+        this.options = options;
+    }
+
+    /**
+     * Obtains value of the specified option by the specified field context.
+     *
+     * @param fieldContext
+     *         the field descriptor to obtain the option
+     * @param option
+     *         the option to obtain
+     * @param <T>
+     *         the type of the option value
+     * @return the {@code Optional} of option value or {@code Optional.empty()}
+     *         if there is no such an option for the field descriptor
+     */
+    public static <T> Optional<T>
+    getOptionValue(FieldContext fieldContext, GeneratedExtension<FieldOptions, T> option) {
```

Comment:
As discussed, let's try to optimize the flow.

1. Find all in-place options and collect validators for them.
2. Find all external constraint options and append the external validators to the collection of validators from p.1.
3. Perform the validation with each of the validators.

Reply:
As discussed vocally, let's check again the current flow after the current PR, while the mentioned optimization is not obvious right now.

Also, the current validation flow relies on by-field validation and making the described change requires reconsidering the whole validation framework underneath functionality.